presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java

```diff
@@ -0,0 +1,261 @@
+                dfTupleDomain.accumulateAndGet(tupleDomain, TupleDomain::intersect);
+                dfState.put(df, DynamicProcessorState.DONE);
+            }
+            else if (jsonResponse.getStatusCode() == Response.Status.NOT_FOUND.getStatusCode()) {
+                dfState.put(df, DynamicProcessorState.NOTFOUND);
+            }
+            else {
+                dfState.put(df, DynamicProcessorState.FAILED);
+            }
+        }
+        catch (Exception exp) {
+            dfState.put(df, DynamicProcessorState.FAILED);
+        }
+    }
+
+    public void collectDynamicSummaryAsync()
+    {
+        CompletableFuture.runAsync(this::collectDynamicSummary, es).whenComplete((value, exception) ->
+        {
+            this.cleanUp();
+        });
+    }
+
+    public void cleanUp()
+    {
+        es.shutdown();
+    }
+
+    private void collectDynamicSummary()
+    {
+        int retry = 0;
+        while (retry < MAX_RETRIES) {
+            if (globalState == DynamicProcessorState.DONE || globalState == DynamicProcessorState.FAILED) {
+                break;
+            }
+            else {
+                collectDynamicSummaryInternal();
+            }
+            retry++;
+        }
+        if (retry == MAX_RETRIES) {
+            afterDFcollection();
+        }
+    }
+
+    private void collectDynamicSummaryInternal()
+    {
+        boolean inProgress = false;
+        for (ConcurrentHashMap.Entry<DynamicFilterExpression, DynamicProcessorState> e : dfState.entrySet()) {
+            DynamicProcessorState state = e.getValue();
+            if (state != DynamicProcessorState.DONE && state != DynamicProcessorState.FAILED) {
+                inProgress = true;
+                fireDFCollection(e.getKey());
+            }
+        }
+        if (!inProgress) {
+            afterDFcollection();
+        }
+    }
+
+    private void afterDFcollection()
+    {
+        if (globalState == DynamicProcessorState.DONE || globalState == DynamicProcessorState.FAILED) {
+            return;
+        }
+        TupleDomain<Symbol> tupleDomain = dfTupleDomain.get();
+        if (!tupleDomain.isAll()) {
+            try {
+                final Expression expr = DomainTranslator.toPredicate(tupleDomain);
+                final Optional<RowExpression> rowExpression = converter.expressionToRowExpression(Optional.of(expr));
+                final RowExpression combinedExpression = call(logicalExpressionSignature(LogicalBinaryExpression.Type.AND),
+                        BOOLEAN,
+                        rowExpression.get(),
+                        staticFilter.get());
+                resultExpression = Optional.of(combinedExpression);
+                globalState = DynamicProcessorState.DONE;
+                triggerCallback(combinedExpression);
+            }
+            catch (Exception e) {
+                globalState = DynamicProcessorState.FAILED;
+                throw e;
```

Comment:
will this exception be reported or logged?

Reply:
no ...but  i added a log for it ... it will be caught upstream though.